### PR TITLE
Mbaas data refator/cache before send

### PR DESCRIPTION
**Pull request by YiMysty**
Please have a look at our [guidelines for contributions](https://github.com/microsoft/appcenter-sdk-android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description
[AB#70996](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/70996)
Refactor to perform local storage write before doing remote operation.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: sdk/appcenter-data/src/androidTest/java/com/microsoft/appcenter/data/LocalDocumentStorageAndroidTest.java [hunk 1]
```diff
@@ -318,6 +318,21 @@ public void updateDocument() {
         assertEquals("Test1", createdDocument.getDeserializedValue());
     }
 
+    @Test
+    public void ReadFailForTransitOperation() {
```
Comment (guperrot) [inline sdk/appcenter-data/src/androidTest/java/com/microsoft/appcenter/data/LocalDocumentStorageAndroidTest.java]:
```suggestion
    public void readFailForTransitOperation() {
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Constants.java [hunk 2]
```diff
@@ -43,6 +43,11 @@
      */
     static final String PENDING_OPERATION_DELETE_VALUE = "DELETE";
 
+    /***
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Constants.java]:
```suggestion
    /**
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 3]
```diff
@@ -375,7 +375,22 @@ public void run() {
...
+                            localDocument);
+                } else {
+                    notifyListenerAndUpdateOperationOnFailure(
+                            new DataException(String.format("Local document with id %s has expired, skip the pending operation processing.", localDocument.getDocumentId())),
+                            localDocument);
+                }
+                mLocalDocumentStorage.deleteOnline(table, localDocument.getPartition(), localDocument.getDocumentId());
+                continue;
+            }
+            System.out.println();
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]:
```suggestion
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 4]
```diff
@@ -375,7 +375,22 @@ public void run() {
     }
 
     private synchronized void processPendingOperations() {
-        for (LocalDocument localDocument : mLocalDocumentStorage.getPendingOperations(Utils.getUserTableName())) {
+        String table = Utils.getUserTableName();
+        for (LocalDocument localDocument : mLocalDocumentStorage.getPendingOperations(table)) {
+            if (localDocument.isExpired()) {
+                if (localDocument.getOperation().equals(Constants.PENDING_OPERATION_PROCESS_VALUE)) {
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: Why are we notifying progress? This will potentially be spammy.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 5]
```diff
@@ -375,7 +375,22 @@ public void run() {
     }
 
     private synchronized void processPendingOperations() {
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]:
```suggestion
    @WorkerThread
    private synchronized void processPendingOperations() {
```

Would need an import though.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 6]
```diff
@@ -561,6 +576,7 @@ public boolean needsRemoteOperation(DocumentWrapper<Void> cachedDocument) {
 
             @Override
             public void callCosmosDb(TokenResult tokenResult, DefaultAppCenterFuture<DocumentWrapper<Void>> result) {
+                mLocalDocumentStorage.deleteOffline(Utils.getTableName(tokenResult.getPartition()), tokenResult.getPartition(), documentId, writeOptions);
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: Accessing storage from main thread is not allowed in strict mode. (HTTP callbacks are executed in main/ui thread).

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 7]
```diff
@@ -859,6 +875,7 @@ public void run() {
 
                                 @Override
                                 public void callCosmosDb(TokenResult tokenResult) {
+                                    mLocalDocumentStorage.createOrUpdateOffline(Utils.getTableName(tokenResult), tokenResult.getPartition(), documentId, document, documentType, writeOptions);
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: same here, need to post

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 8]
```diff
@@ -96,6 +98,15 @@
     private static final String BY_PARTITION_AND_DOCUMENT_ID_WHERE_CLAUSE =
             String.format("%s = ? AND %s = ?", PARTITION_COLUMN_NAME, DOCUMENT_ID_COLUMN_NAME);
 
+    /**
+     * String columns can be updated.
+     */
+    private static final String[] MUTABLE_STRING_COLUMNS = new String[]
+            {DOCUMENT_COLUMN_NAME, ETAG_COLUMN_NAME, PENDING_OPERATION_COLUMN_NAME};
+
+    private static final String[] MUTABLE_LONG_COLUMNS = new String[]
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: As in long columns cannot be updated?

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 4 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: it's a special condition :) 
This case happens when user set the writeOptions to no cache:
in this case if the pending operation failed in future, we cannot redo anything, thus we need to notify the operation failed :D and this failed is not like set the writeOptions to INFINITE.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java (hunk 8 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: ah they can be updated... 🤔 is it a bad naming problem?

Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: this one does have the comment

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 6 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: hmm I got the point, but the tricky part for this is we need to know what is the table name to write.... And the way we get the table name is to parse the token result from the token-exchange service.
And we can use the cache token for sure ":), but.... the first time will not have cache... So not sure how to refactory it...

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 7 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: comment above~

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java (hunk 8 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: I guess you miss a `not` here 😅 done~

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 6 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: changed to async, not sure if correct :D

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 4 above)
Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: I don't understand how it's related to the new constant, and why it's checked only when network comes back.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 9]
```diff
@@ -867,16 +873,24 @@ public void onCallFailed(Exception e) {
             @Override
             public void run() {
                 if (mNetworkStateHelper.isNetworkConnected()) {
+
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]:
```suggestion
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 10]
```diff
@@ -1037,7 +1068,7 @@ public void run() {
                                     eTag),
                             null);
                 }
-                if (pendingOperation.isExpired() || PENDING_OPERATION_DELETE_VALUE.equals(pendingOperation.getOperation())) {
+                if (PENDING_OPERATION_DELETE_VALUE.equals(pendingOperation.getOperation())) {
 
                     /* Remove the document if expiration_time has elapsed or it is a delete operation. */
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: Comment out of date when removing one of the condition (also why is this change needed?).

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 11]
```diff
@@ -96,6 +98,18 @@
     private static final String BY_PARTITION_AND_DOCUMENT_ID_WHERE_CLAUSE =
             String.format("%s = ? AND %s = ?", PARTITION_COLUMN_NAME, DOCUMENT_ID_COLUMN_NAME);
 
+    /**
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: Can you explain the changes on that class? It's hard to read the changes and I don't understand what it's supposed to be doing now.

Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: these are for supporting another method called `getContentValues`, because another method need to pass everything... which is not accommodate to the situation... (for some column like download time, which requires first read the document and then write the value again to keep unchanged... which seems too redundant... :D )

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 10 above)
Comment (YiMysty) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: the pendOperation expiration check will happened in the processPendingOperation now, that's why removed from here :D

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 12]
```diff
@@ -354,6 +369,47 @@ private static ContentValues getContentValues(
...
+
+                /* Successfully create or update documents online, remove the pending operation from  localstorage. **/
+                return mDatabaseManager.delete(table, document.getId());
+            } else {
+
+                /*
+                  The pending operation is delete/create/update with writeOptions no cache. It will be execute:
+                  1. Before every delete/create/update call to cosmosdb.
+                 */
+                values = getContentValues(document.getPartition(), document.getId(), new HashMap<String, String>() {{
```
Comment (YiMysty) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: this is incorrect need to mark the value to null... forget about that :(

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 4 above)
Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: Discussed offline.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 13]
```diff
@@ -374,8 +374,23 @@ public void run() {
         }
     }
 
+    @WorkerThread
     private synchronized void processPendingOperations() {
-        for (LocalDocument localDocument : mLocalDocumentStorage.getPendingOperations(Utils.getUserTableName())) {
+        String table = Utils.getUserTableName();
+        for (LocalDocument localDocument : mLocalDocumentStorage.getPendingOperations(table)) {
+            if (localDocument.isExpired()) {
+                if (localDocument.getOperation().equals(Constants.PENDING_OPERATION_PROCESS_VALUE)) {
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: can we test that if first and not nest it under `if (localDocument.isExpired())`? It's very confusing otherwise.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 14]
```diff
@@ -161,23 +175,24 @@ private static SQLiteQueryBuilder getPartitionAndDocumentIdQueryBuilder() {
         return builder;
     }
 
-    private <T> long write(String table, DocumentWrapper<T> document, WriteOptions writeOptions, String pendingOperationValue) {
-        if (writeOptions.getDeviceTimeToLive() == TimeToLive.NO_CACHE) {
-            return 0;
+    private static ContentValues getContentValues(
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: https://github.com/microsoft/appcenter-sdk-android/pull/1289#discussion_r337283156 should be summarized as a Javadoc method comment here.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 15]
```diff
@@ -354,6 +369,48 @@ private static ContentValues getContentValues(
...
+            return mDatabaseManager.replace(table, values, PARTITION_COLUMN_NAME, DOCUMENT_ID_COLUMN_NAME);
+        } else {
+
+            /*
+              The operation is called after successfully/failed to call cosmosdb api to do create/update and writeOptions is no cache, remove the trace log.
+             */
+            if (pendingOperationValue == null) {
+                AppCenterLog.debug(LOG_TAG, String.format("Trying to remove %s:%s document trace from cache", document.getPartition(), document.getId()));
+
+                /* Successfully create or update documents online, remove the pending operation from  localstorage. **/
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
                /* Successfully create or update documents online, remove the pending operation from the local storage. */
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 16]
```diff
@@ -393,8 +450,18 @@ void updatePendingOperation(LocalDocument operation) {
         values = mDatabaseManager.nextValues(cursor);
         cursor.close();
         if (values != null) {
-            if (ReadOptions.isExpired(values.getAsLong(EXPIRATION_TIME_COLUMN_NAME))) {
-                mDatabaseManager.delete(table, values.getAsLong(DatabaseManager.PRIMARY_KEY));
+            long documentExpirationTime = values.getAsLong(EXPIRATION_TIME_COLUMN_NAME);
+            String pendingOperation = values.getAsString(PENDING_OPERATION_COLUMN_NAME);
+
+            /*  This happens when doing operation cosmosdb create/update when writeOptions in flight and did not resolve a result yet. **/
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
            /* This happens when doing operation cosmosdb create/update when writeOptions in flight and did not resolve a result yet. */
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java (hunk 14 above)
Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: Also move it next to the other getContentValues.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Constants.java [hunk 17]
```diff
@@ -43,6 +43,11 @@
      */
     static final String PENDING_OPERATION_DELETE_VALUE = "DELETE";
 
+    /**
+     * Generic pending operation when writeOptions is no cache.
+     */
+    static final String PENDING_OPERATION_PROCESS_VALUE = "PROCESSING";
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Constants.java]: PROCESSING in the constant then

Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Constants.java]: Bump here.

**Comment by guperrot** [comment]
Please test online and offline in the test app and report back.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 18]
```diff
@@ -374,8 +374,24 @@ public void run() {
...
     }
 
+    @WorkerThread
     private synchronized void processPendingOperations() {
-        for (LocalDocument localDocument : mLocalDocumentStorage.getPendingOperations(Utils.getUserTableName())) {
+        String table = Utils.getUserTableName();
+        for (LocalDocument localDocument : mLocalDocumentStorage.getPendingOperations(table)) {
+            if (localDocument.getOperation().equals(Constants.PENDING_OPERATION_PROCESS_VALUE)) {
+                notifyListenerAndUpdateOperationOnFailure(
+                        new DataException(String.format("Remote state is unknown for document %s, and no local cache for this document.", localDocument.getDocumentId())),
```
Comment (jwallra) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]:
```suggestion
                        new DataException(String.format("Remote state is unknown for document id %s, and no local cache exists for this document.", localDocument.getDocumentId())),
```

**Comment by YiMysty** [comment]
> No recent activity on this so ping to make sure.
> Here is the list of unresolved feedbacks:
> 
> [#1289 (comment)](https://github.com/microsoft/appcenter-sdk-android/pull/1289#issuecomment-545085686)
> [#1289 (comment)](https://github.com/microsoft/appcenter-sdk-android/pull/1289#discussion_r337299408)
> [#1289 (comment)](https://github.com/microsoft/appcenter-sdk-android/pull/1289#discussion_r337300187)
> [#1289 (comment)](https://github.com/microsoft/appcenter-sdk-android/pull/1289#discussion_r337300291)

sorry shift draft to another task... will look it latter 🙈
